lib/src/stream_extensions.dart

```diff
@@ -33,4 +33,21 @@ extension StreamExtensions<T> on Stream<T> {
       sink.close();
     }));
   }
+
+  /// Returns the first data or error event emitted by [this] if it emits any,
+  /// or `null` if it emits a done event first.
+  Future<T?> get firstOrNull {
+    var completer = Completer<T?>.sync();
+    late StreamSubscription<T> subscription;
+    subscription = listen((event) {
+      subscription.cancel();
+      completer.complete(event);
+    }, onError: (Object error, StackTrace stackTrace) {
+      subscription.cancel();
+      completer.completeError(error, stackTrace);
+    }, onDone: () {
+      completer.complete(null);
+    });
```

Comment:
For wherever we end up landing this...

I think we can avoid the `late`

```suggestion
    final subscription = listen(null);
    subscription
      ..onData((event) {
        subscription.cancel();
        completer.complete(event);
      })
      ..onError((Object error, StackTrace stackTrace) {
        subscription.cancel();
        completer.completeError(error, stackTrace);
      })
      ..onDone(() {
        completer.complete(null);
      });
```